src/aes128gcm.rs

```diff
@@ -23,174 +28,380 @@ const ECE_WEBPUSH_IKM_LENGTH: usize = 32;
-        local_prv_key: &dyn LocalKeyPair,
-        remote_pub_key: &dyn RemotePublicKey,
-        auth_secret: &[u8],
-        plaintext: &[u8],
-        params: WebPushParams,
-    ) -> Result<Vec<u8>> {
-        let cryptographer = crypto::holder::get_cryptographer();
-        let salt = match params.salt {
-            Some(salt) => salt,
-            None => {
-                let mut salt = [0u8; ECE_SALT_LENGTH];
-                cryptographer.random_bytes(&mut salt)?;
-                salt.to_vec()
-            }
-        };
-        let mut header = vec![0u8; ECE_AES128GCM_HEADER_LENGTH + ECE_WEBPUSH_PUBLIC_KEY_LENGTH];
-        header[0..ECE_SALT_LENGTH].copy_from_slice(&salt);
-        BigEndian::write_u32(&mut header[ECE_SALT_LENGTH..], params.rs);
-        header[ECE_SALT_LENGTH + 4] = ECE_WEBPUSH_PUBLIC_KEY_LENGTH as u8;
-        let raw_local_pub_key = local_prv_key.pub_as_raw()?;
-        header[ECE_AES128GCM_HEADER_LENGTH
-            ..ECE_AES128GCM_HEADER_LENGTH + ECE_WEBPUSH_PUBLIC_KEY_LENGTH]
-            .copy_from_slice(&raw_local_pub_key);
-        let mut ciphertext = Self::common_encrypt(
-            local_prv_key,
-            remote_pub_key,
-            auth_secret,
-            &salt,
-            params.rs,
-            params.pad_length,
-            plaintext,
+pub(crate) fn decrypt(
+    local_prv_key: &dyn LocalKeyPair,
+    auth_secret: &[u8],
+    ciphertext: &[u8],
+) -> Result<Vec<u8>> {
+    let cryptographer = crypto::holder::get_cryptographer();
+    if ciphertext.is_empty() {
+        return Err(Error::ZeroCiphertext);
+    }
+
+    // Buffer into which to write the output.
+    // This will avoid any reallocations because plaintext will always be smaller than ciphertext.
+    // We could calculate a tighter bound if memory usage is an issue in future.
+    let mut output = Vec::<u8>::with_capacity(ciphertext.len());
+
+    let header = Header::read_from(ciphertext)?;
+    if ciphertext.len() == header.encoded_size() {
+        return Err(Error::ZeroCiphertext);
+    }
+
+    // The `keyid` field must contain the serialized ephemeral public key.
+    if header.keyid.len() != ECE_WEBPUSH_PUBLIC_KEY_LENGTH {
+        return Err(Error::InvalidKeyLength);
+    }
+    let remote_pub_key = cryptographer.import_public_key(&header.keyid)?;
+
+    let (key, nonce) = derive_key_and_nonce(
+        cryptographer,
+        EceMode::DECRYPT,
+        local_prv_key,
+        &*remote_pub_key,
+        auth_secret,
+        header.salt,
+    )?;
+
+    // We'll re-use this buffer as scratch space for decrypting each record.
+    // This is nice for memory usage, but actually the main motivation is that have the decryption
+    // output a `PlaintextRecord` struct, which holds a borrowed slice of plaintext.
+    let mut plaintext_buffer = vec![0u8; (header.rs as usize) - ECE_TAG_LENGTH];
+
+    let records = ciphertext[header.encoded_size()..].chunks(header.rs as usize);
+
+    let mut seen_final_record = false;
```

Comment:
My concern here specifically is compatibility with potential large send-tab payloads sent from desktop, which defaults to `rs=4096` and might split large payloads into several records.

Reply:
Ah, so those records would come in via alternate methods. Still feels like this is more a bandage sort of thing and we're covering up a problem. Granted, even if we were to enforce a one block limit, i suppose we'd have to have some sort of work-around for existing records that might be in transition. Just feels a bit odd.

Reply:
It's also worth nothing that this shouldn't really be observable through the public API of the crate - if you call `ece::encrypt` with a payload larger than 4096 bytes, it will expand the record size in order to accommodate it and encrypt it all in a single record.

That said, I do intend to bring back multi-record support in a follow-up PR, it just seemed like a bit much to take on for review all at once here! If you think it's worthwhile, I can hold off on merging this until after I've attempted that work and restored the symmetry between encrypting and decrypting.

Reply:
Heh, if I had a real problem with it, I wouldn't have approved. Mostly just noting and debating if it something we might clean up in the future. Sounds like you're several steps ahead of me.

again.

Still 👍🏻